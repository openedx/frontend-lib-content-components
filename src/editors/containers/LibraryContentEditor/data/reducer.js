import { createSlice } from '@reduxjs/toolkit';
import { modes } from '../constants';
import { StrictDict } from '../../../utils';

const initialState = {
  libraries: {},
  selectedLibraryId: null,
  savedLibraryId: null,
  settings: {
    // [libraryId]: initialLibrarySettings,
    //  This reducer structure allows selected settings
    //  to persist when user switches between libraries.
  },
};

const initialLibrarySettings = {
  version: null,
  mode: modes.random.value,
  count: -1,
  showReset: false,
  blocks: [],               // [ { id: 'id', display_name: 'name', block_type: 'type' } ]
  candidates: [],           // ['id1', 'id2']
  v1BlockRequests: {},      // { [blockId]: 'RequestkeyState' }
};

const library = createSlice({
  name: 'library',
  initialState,
  reducers: {
    initializeFromBlockValue: (state, { payload }) => ({
      ...state,
      savedLibraryId: payload.libraryId,
      selectedLibraryId: payload.libraryId,
      settings: payload.settings,
    }),
    addLibraries: (state, { payload }) => ({
      ...state,
      libraries: {
        ...state.libraries,
        ...payload.libraries,
      },
    }),
    initialLibrarySettings: (state, { payload }) => ({
      ...state,
      settings: {
        ...state.settings,
        [payload.selectedLibraryId]: initialLibrarySettings,
      },
    }),
    setLibraryId: (state, { payload }) => ({
      ...state,
      selectedLibraryId: payload.selectedLibraryId,
    }),
    setLibraryVersion: (state, { payload }) => ({
      ...state,
      settings: {
        ...state.settings,
        [state.selectedLibraryId]: {
          ...state.settings[state.selectedLibraryId],
          version: payload.version,
        },
      },
    }),
    setModeForLibrary: (state, { payload }) => ({
      ...state,
      settings: {
        ...state.settings,
        [state.selectedLibraryId]: {
          ...state.settings[state.selectedLibraryId],
          mode: payload.mode,
        },
      },
    }),
    setCountForLibrary: (state, { payload }) => ({
      ...state,
      settings: {
        ...state.settings,
        [state.selectedLibraryId]: {
          ...state.settings[state.selectedLibraryId],
          count: payload.count,
        },
      },
    }),
    setShowResetForLibrary: (state, { payload }) => ({
      ...state,
      settings: {
        ...state.settings,
        [state.selectedLibraryId]: {
          ...state.settings[state.selectedLibraryId],
          showReset: payload.showReset,
        },
      },
    }),
    setLibraryBlocks: (state, { payload }) => ({
      ...state,
      settings: {
        ...state.settings,
        [state.selectedLibraryId]: {
          ...state.settings[state.selectedLibraryId],
          blocks: payload.blocks,
        },
      },
    }),
    setCandidatesForLibrary: (state, { payload }) => ({
      ...state,
      settings: {
        ...state.settings,
        [state.selectedLibraryId]: {
          ...state.settings[state.selectedLibraryId],
          candidates: payload.candidates,
        },
      },
    }),
    setV1BlockRequests: (state, { payload }) => ({
      ...state,
      settings: {
        ...state.settings,
        [state.selectedLibraryId]: {
          ...state.settings[state.selectedLibraryId],
          v1BlockRequests: payload.v1BlockRequests,
        },
      },
    }),
    updateV1BlockRequestStatus: (state, { payload }) => ({
      ...state,
      settings: {
        ...state.settings,
        [state.selectedLibraryId]: {
          ...state.settings[state.selectedLibraryId],
          v1BlockRequests: {
            ...state.settings[state.selectedLibraryId].v1BlockRequests,
            [payload.blockId]: payload.status,
          },
        },
      },
    }),
    addLibraryBlock: (state, { payload }) => ({
      ...state,
      settings: {
        ...state.settings,
        [state.selectedLibraryId]: {
          ...state.settings[state.selectedLibraryId],
          blocks: [
            ...state.settings[state.selectedLibraryId].blocks,
            payload.block,
          ],
        },
      },
    }),
  },
});

const actions = StrictDict(library.actions);

const { reducer } = library;

export {
  actions,
  initialState,
  initialLibrarySettings,
  reducer,
};
