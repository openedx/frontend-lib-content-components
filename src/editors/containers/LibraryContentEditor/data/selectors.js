import { createSelector } from 'reselect';
import { modes } from '../constants';
import { getUsageKey } from '../utils';
import * as module from './selectors';

export const libraryState = (state) => state.library;

const mkSimpleSelector = (cb) => createSelector([module.libraryState], cb);

export const simpleSelectors = {
  libraries: mkSimpleSelector(library => library.libraries),
  savedLibraryId: mkSimpleSelector(library => library.savedLibraryId),
  selectedLibraryId: mkSimpleSelector(library => library.selectedLibraryId),
  settings: mkSimpleSelector(library => library.settings),
};

export const blocks = createSelector(
  [
    module.simpleSelectors.selectedLibraryId,
    module.simpleSelectors.settings,
  ],
  (
    selectedLibraryId,
    settings,
  ) => {
    if (selectedLibraryId) {
      return settings[selectedLibraryId]?.blocks ?? [];
    }
    return [];
  },
);

export const v1BlockRequests = createSelector(
  [
    module.simpleSelectors.selectedLibraryId,
    module.simpleSelectors.settings,
  ],
  (
    selectedLibraryId,
    settings,
  ) => {
    if (selectedLibraryId) {
      return settings[selectedLibraryId]?.v1BlockRequests ?? {};
    }
    return [];
  },
);

export const candidates = createSelector(
  [
    module.simpleSelectors.selectedLibraryId,
    module.simpleSelectors.settings,
  ],
  (
    selectedLibraryId,
    settings,
  ) => {
    if (selectedLibraryId) {
      return settings[selectedLibraryId]?.candidates ?? [];
    }
    return [];
  },
);

export const mode = createSelector(
  [
    module.simpleSelectors.selectedLibraryId,
    module.simpleSelectors.settings,
  ],
  (
    selectedLibraryId,
    settings,
  ) => {
    if (selectedLibraryId) {
      return settings[selectedLibraryId]?.mode ?? modes.random.value;
    }
    return modes.random.value;
  },
);

export const libraryPayload = createSelector(
  [
    module.simpleSelectors.selectedLibraryId,
    module.simpleSelectors.settings,
  ],
  (
    selectedLibraryId,
    settings,
  ) => {
    let manual = false;
    let shuffle = true;
    let count = null;
    if (selectedLibraryId && settings[selectedLibraryId]) {
      count = settings[selectedLibraryId].count ?? 1;
      if (settings[selectedLibraryId].mode === modes.selected.value) {
        manual = true;
        shuffle = false;
        count = -1;
      }
    }
    return {
      libraryId: selectedLibraryId,
      libraryVersion: settings[selectedLibraryId]?.version,
      manual,
      shuffle,
      count,
      showReset: settings[selectedLibraryId]?.showReset,
      candidates: settings[selectedLibraryId]?.candidates || [],
    };
  },
);

export default {
  ...simpleSelectors,
  blocks,
  candidates,
  mode,
  libraryPayload,
  v1BlockRequests,
};
