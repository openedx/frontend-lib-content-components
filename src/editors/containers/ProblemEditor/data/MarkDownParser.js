/*
This is the direct motivation code taken from:
https://github.com/open-craft/capa-visual-editor/

Changes have been made to adapt the parser with respect to our use case.
*/

import { ProblemTypeKeys } from "../../../data/constants/problem";

export const indexToLetterMap = [...Array(26)].map((val, i) => String.fromCharCode(i + 65));

function getHints(markdown) {
  if (!markdown || !markdown.trim()) {
    return [];
  }
  const hints = [];

  function getHint(row) {
    let hint = '';
    if (row.startsWith('||')) {
      const hintStart = row.indexOf('||');
      const hintEnd = row.slice(hintStart + 2).indexOf('||');
      hint = row.slice(hintStart + 2, hintEnd + 2);
    }
    return hint;
  }
  for (const d in markdown.split('\n')) {
    const row = markdown.split('\n')[d];
    const hint = getHint(row);
    if (hint.length) {
      hints.push({
        id: Number(d),
        value: hint,
      });
    }
  }
  return hints;
}

function getShortAnswerOptions(markdown) {
  // TODO: Include tolerence for number it fails for now.
  let problemType = ProblemTypeKeys.TEXTINPUT;
    if ((!markdown || !markdown.trim())) {
      return {
        shortAnswersList: [],
        problemType,
      };
    }
    const markdownListData = markdown.split('\n');
    const shortAnswersList = [];
    let gotAnswer = false;
    for (const i in markdownListData) {
      const row = markdownListData[i].trim();
      const feedbackStart = row.indexOf('{{');
      const feedbackEnd = row.indexOf('}}');
      let feedback = '';
      let answer = '';
      let correct = true;
      if (row.startsWith('=') && !gotAnswer) {
        gotAnswer = true;
        if (feedbackStart > -1) {
          feedback = row.slice(feedbackStart + 2, feedbackEnd);
          answer = row.slice(1, feedbackStart).trim();
        } else {
          answer = row.slice(1).trim();
        }
      } else if (row.startsWith('=') && gotAnswer) {
        break;
      } else if (gotAnswer && row.startsWith('or=')) { // the next correct answer
        if (feedbackStart > -1) {
          answer = row.slice(3, feedbackStart).trim();
          feedback = row.slice(feedbackStart + 2, feedbackEnd);
        } else {
          answer = row.slice(3).trim();
        }
      } else if (gotAnswer && row.startsWith('not=')) { // the incorrect answer
        if (feedbackStart > -1) {
          answer = row.slice(4, feedbackStart).trim();
          feedback = row.slice(feedbackStart + 2, feedbackEnd);
        } else {
          answer = row.slice(4).trim();
        }
        correct = false;
      }
      if (answer) {
        problemType = isFinite(answer) ? ProblemTypeKeys.NUMERIC : ProblemTypeKeys.TEXTINPUT;
        shortAnswersList.push({
          id: indexToLetterMap[shortAnswersList.length],
          title: answer,
          feedback,
          correct,
        });
      }
    }
    return {
      shortAnswersList,
      problemType,
    };
}

function getMultipleChoiceOptions(markdown) {
  const multipleChoiceOptions = [];
  const data = {
    multiSelectAnswersList: multipleChoiceOptions,
    problemType: ProblemTypeKeys.MULTISELECT,
    groupFeedbackList: [],
  };
  if (!markdown || !markdown.trim()) {
    return data;
  }
  const markdownListData = markdown.split('\n');
  let feedback = '';
  for (const d in markdownListData) {
    const row = markdownListData[d];
    // multiple choices
    if (row.startsWith('[ ]') || row.startsWith('[x]')) {
      // title
      let title = row.slice(row.indexOf(']') + 1).trim();

      const feedbacksStart = row.indexOf('{{');
      const selectedFeedbackStart = row.indexOf(' selected:') || row.indexOf('{selected:'); // two ways to define selected feedback
      const selectedFeedbackEnd = row.slice(selectedFeedbackStart + 1).indexOf('}') + selectedFeedbackStart;
      const unselectedFeedbackStart = row.indexOf('unselected:');
      const unselectedFeedbackEnd = row.slice(unselectedFeedbackStart + 1).indexOf('}}') + unselectedFeedbackStart;

      let selectedFeedback = '';
      let unselectedFeedback = '';
      let titleChanged = false;
      if (feedbacksStart !== -1 && selectedFeedbackStart !== -1) {
        selectedFeedback = row.slice(selectedFeedbackStart + 10, selectedFeedbackEnd);
        title = row.slice(row.indexOf(']') + 1, feedbacksStart).trim();
        titleChanged = true;
      }
      if (feedbacksStart !== -1 && unselectedFeedbackStart !== -1) {
        unselectedFeedback = row.slice(unselectedFeedbackStart + 12, unselectedFeedbackEnd + 1);
        if (!titleChanged) {
          title = row.slice(row.indexOf(']') + 1, feedbacksStart).trim();
        }
      }
      multipleChoiceOptions.push({
        id: indexToLetterMap[multipleChoiceOptions.length],
        title,
        correct: !row.startsWith('[ ]'),
        selectedFeedback,
        unselectedFeedback,
      });
    }
    // group Feedback
    if (multipleChoiceOptions.length) {
      if (row.startsWith('{{')) {
        const groupFeedbackEnd = row.indexOf('}}');
        const goupStart = row.indexOf('((');
        const groupEnd = row.indexOf('))');
        const group = row.slice(goupStart + 2, groupEnd).split(' ');
        const groupChoices = [];
        for (const i in group) {
          const letter = group[i];
          if (multipleChoiceOptions[indexToLetterMap.indexOf(letter)]) {
            groupChoices.push(multipleChoiceOptions[indexToLetterMap.indexOf(letter)].id);
          }
        }
        feedback = row.slice(groupEnd + 2, groupFeedbackEnd).trim();

        if (groupChoices.length && feedback) {
          data.groupFeedbackList.push({
            id: groupChoices.length,
            answers: groupChoices,
            feedback,
          });
        }
      }
    }
  }
  return data;
}

function getSingleChoiceOptions(markdown) {
  // TODO: Build support for selected and unselected feedback
  if (!markdown || !markdown.trim()) {
    return {
      singleSelectAnswersList: [],
      problemType: ProblemTypeKeys.DROPDOWN,
    };
  }

  const singleChoiceOptions = [];
  let dropDownMode = false;
  const markdownListData = markdown.split('\n');
  for (const d in markdownListData) {
    const row = markdownListData[d].trim();
    if (row.startsWith('[[')) {
      dropDownMode = true;
      continue;
    }
    if (row.startsWith(']]')) {
      break;
    }
    if (dropDownMode) {
      if (row.startsWith('(')) {
        let feedback = '';
        let title = row.slice(row.indexOf('(') + 1, row.indexOf(')')).trim();
        const feedbackStart = row.indexOf('{{');
        const feedbackEnd = row.indexOf('}}');
        if (feedbackStart !== -1 && feedbackEnd !== -1 && feedbackEnd > feedbackStart) { // assum the row contains feedback
          feedback = row.slice(feedbackStart + 2, feedbackEnd).trim();
          title = title.slice(0, feedbackStart).trim();
        }
        const correct = true;
        singleChoiceOptions.push({
          id: indexToLetterMap[singleChoiceOptions.length],
          title,
          correct,
          feedback,
        });
      } else {
        let feedback = '';
        let title = '';
        // let title = row.slice(row.indexOf(')') + 1).trim();

        const feedbackStart = row.indexOf('{{');
        const feedbackEnd = row.indexOf('}}');
        if (feedbackStart !== -1 && feedbackEnd !== -1 && feedbackEnd > feedbackStart) { // assum the row contains feedback
          feedback = row.slice(feedbackStart + 2, feedbackEnd).trim();
          title = row.slice(0, feedbackStart).trim();
        } else {
          title = row.trim();
        }
        const correct = false;
        singleChoiceOptions.push({
          id: indexToLetterMap[singleChoiceOptions.length],
          title,
          correct,
          feedback,
        });
      }
    } else if (row.startsWith('( )') || row.startsWith('(x)')) {
      let feedback = '';
      let title = row.slice(row.indexOf(')') + 1).trim();
      const feedbackStart = row.indexOf('{{');
      const feedbackEnd = row.indexOf('}}');
      if (feedbackStart !== -1 && feedbackEnd !== -1 && feedbackEnd > feedbackStart) { // assum the row contains feedback
        feedback = row.slice(feedbackStart + 2, feedbackEnd).trim();
        title = title.slice(0, feedbackStart - 4).trim();
      }
      const correct = !!row.startsWith('(x)');
      singleChoiceOptions.push({
        id: indexToLetterMap[singleChoiceOptions.length],
        title,
        correct,
        feedback,
      });
    }
  }

  return {
    singleSelectAnswersList: singleChoiceOptions,
    problemType: dropDownMode ? ProblemTypeKeys.DROPDOWN : ProblemTypeKeys.SINGLESELECT,
  };
}

function getEditorData(markdown) {
  // TODO: Prase heading in this function.
  if (!markdown || !markdown.trim()) {
    return '';
  }
  let description = '';
  const markdownData = markdown.trim().split('\n');
  for (const i in markdownData) {
    const row = markdownData[i];
    if (['{', '(', '[', '=', '|'].indexOf(row.trim()[0]) === -1) {
      const questionTextStart = row.indexOf('>>');
      const questionTextEnd = row.indexOf('<<');
      let questionText = row;
      if (questionTextStart !== -1 && questionTextEnd !== -1 && (questionTextEnd > questionTextStart)) {
        questionText = row.replace('>>', '<p>').replace('<<', '</p>');
      }
      description += `${questionText}\n`;
    } else {
      break;
    }
  }
  return description;
}

function parseMarkdown(markdown) {
  if (!markdown || !markdown.trim()) {
    return {};
  }
  let answers = {};
  let problemType = '';
  const question = getEditorData(markdown);
  const hints = getHints(markdown);
  let data = {question, settings: {hints}};
  let parsedData = getSingleChoiceOptions(markdown);
  if (parsedData.singleSelectAnswersList.length !== 0){
    answers = parsedData.singleSelectAnswersList;
    problemType = parsedData.problemType;
    data = {...data, answers, problemType};
  }
  parsedData = getMultipleChoiceOptions(markdown);
  if (parsedData.multiSelectAnswersList.length !== 0){
    answers = parsedData.multiSelectAnswersList;
    problemType = parsedData.problemType;
    const groupFeedbackList = parsedData.groupFeedbackList;
    data = {...data, answers, problemType, groupFeedbackList};
  }
  parsedData = getShortAnswerOptions(markdown);
  if (parsedData.shortAnswersList.length !== 0){
    answers = parsedData.shortAnswersList;
    problemType = parsedData.problemType;
    data = {...data, answers, problemType};
  }
  return data;
}

function getScorringSettings() {
  return {
    selectedAttemptsOption: window.LXCData.max_attempts,
    selectedPointOption: window.LXCData.weight,
  };
}

export {
  getMultipleChoiceOptions,
  getSingleChoiceOptions,
  getEditorData,
  getHints,
  getScorringSettings,
  getShortAnswerOptions,
  parseMarkdown,
};
