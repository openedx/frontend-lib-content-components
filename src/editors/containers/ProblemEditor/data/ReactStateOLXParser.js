import _ from 'lodash-es';
import { XMLBuilder } from 'fast-xml-parser';
import { ProblemTypeKeys } from '../../../data/constants/problem';

class ReactStateOLXParser {
  constructor(problemState) {
    this.parserOptions = {
      ignoreAttributes: false,
      alwaysCreateTextNode: true,
    };
    this.builderOptions = {
      ignoreAttributes: false,
      attributeNamePrefix: '@_',
      suppressBooleanAttributes: false,
      format: true,
    };
    this.builder = new XMLBuilder(this.builderOptions);
    this.problemState = problemState.problem;
  }

  addHints() {
    const hintsArray = [];
    const hints = _.get(this.problemState, 'settings.hints', []);
    hints.forEach(element => {
      hintsArray.push({
        '#text': element.value,
      });
    });
    const demandhint = {
      demandhint: {
        hint: hintsArray,
      },
    };
    return demandhint;
  }

  addMultiSelectAnswers(option) {
    const choice = [];
    let compoundhint = [];
    let widget = {};
    const { answers } = this.problemState;
    answers.forEach((answer) => {
      const feedback = [];
      let singleAnswer = {};
      if (this.hasAttributeWithValue(answer, 'title')) {
        if (this.hasAttributeWithValue(answer, 'selectedFeedback')) {
          feedback.push({
            '#text': _.get(answer, 'selectedFeedback'),
            '@_selected': true,
          });
        }
        if (this.hasAttributeWithValue(answer, 'unselectedFeedback')) {
          feedback.push({
            '#text': _.get(answer, 'unselectedFeedback'),
            '@_selected': false,
          });
        }
        if (this.hasAttributeWithValue(answer, 'feedback')) {
          feedback.push({
            '#text': _.get(answer, 'feedback'),
          });
        }
        if (feedback.length) {
          singleAnswer[`${option}hint`] = feedback;
        }
        singleAnswer = {
          '#text': answer.title,
          '@_correct': answer.correct,
          ...singleAnswer,
        };
        choice.push(singleAnswer);
      }
    });
    widget = { [option]: choice };
    if (_.has(this.problemState, 'groupFeedbackList')) {
      compoundhint = this.addGroupFeedbackList();
      widget = {
        ...widget,
        compoundhint,
      };
    }
    return widget;
  }

  addGroupFeedbackList() {
    const compoundhint = [];
    const { groupFeedbackList } = this.problemState;
    groupFeedbackList.forEach((element) => {
      compoundhint.push({
        '#text': element.feedback,
        '@_value': element.answers.join(' '),
      });
    });
    return compoundhint;
  }

  buildOLXWithQuestion(problemObject, problemType) {
    const olxWithoutQuestion = this.builder.build(problemObject);

    return this.insertQuestionIntoOLX(olxWithoutQuestion, problemType);
  }

  insertQuestionIntoOLX(olx, tagName) {
    const { question } = this.problemState;

    const tagPattern = new RegExp(`<${tagName}[^>]*>`);
    const match = olx.match(tagPattern);
    const insertIndex = match.index + match[0].length;

    const olxWithQuestion = `${olx.slice(0, insertIndex)}\n${question}${olx.slice(insertIndex)}`;

    return olxWithQuestion;
  }

  buildMultiSelectProblem(problemType, widget, option) {
    const widgetObject = this.addMultiSelectAnswers(option);
    const demandhint = this.addHints();
    const problemObject = {
      problem: {
        [problemType]: {
          [widget]: widgetObject,
        },
        ...demandhint,
      },
    };

    return this.buildOLXWithQuestion(problemObject, problemType);
  }

  buildTextInput() {
    const problemType = ProblemTypeKeys.TEXTINPUT;
    const demandhint = this.addHints();
    const answerObject = this.buildTextInputAnswersFeedback();
    const problemObject = {
      problem: {
        [problemType]: {
          ...answerObject,
        },
        ...demandhint,
      },
    };

    return this.buildOLXWithQuestion(problemObject, problemType);
  }

  buildTextInputAnswersFeedback() {
    const { answers } = this.problemState;
    let answerObject = {};
    const additionAnswers = [];
    const wrongAnswers = [];
    let firstCorrectAnswerParsed = false;
    answers.forEach((answer) => {
      const correcthint = this.getAnswerHints(answer);
      if (this.hasAttributeWithValue(answer, 'title')) {
        if (answer.correct && firstCorrectAnswerParsed) {
          additionAnswers.push({
            '@_answer': answer.title,
            ...correcthint,
          });
        } else if (answer.correct && !firstCorrectAnswerParsed) {
          firstCorrectAnswerParsed = true;
          answerObject = {
            '@_answer': answer.title,
            ...correcthint,
          };
        } else if (!answer.correct) {
          wrongAnswers.push({
            '@_answer': answer.title,
            '#text': answer.feedback,
          });
        }
      }
    });
    answerObject = {
      ...answerObject,
      additional_answer: additionAnswers,
      stringequalhint: wrongAnswers,
      '@_type': _.get(this.problemState, 'additionalAttributes.type', 'ci'),
      textline: {
        '@_size': _.get(this.problemState, 'additionalAttributes.textline.size', 20),
      },
    };
    return answerObject;
  }

  buildNumericInput() {
    const demandhint = this.addHints();
    const answerObject = this.buildNumericalResponse();
    const problemObject = {
      problem: {
        [ProblemTypeKeys.NUMERIC]: answerObject,
        ...demandhint,
      },
    };

    // Passing 'problem' as problem type, because for this type of problems the question has to be inserted under
    // <problem> tag, instead of <${problemType}> tag.
    return this.buildOLXWithQuestion(problemObject, 'problem');
  }

  buildNumericalResponse() {
    const { answers } = this.problemState;
    let answerObject = {};
    const additionalAnswers = [];
    let firstCorrectAnswerParsed = false;
    /*
    TODO: Need to figure out how to add multiple numericalresponse,
    the parser right now converts all the other right answers into
    additional answers.
    */
    answers.forEach((answer) => {
      const correcthint = this.getAnswerHints(answer);
      if (this.hasAttributeWithValue(answer, 'title')) {
        if (answer.correct && !firstCorrectAnswerParsed) {
          firstCorrectAnswerParsed = true;
          let responseParam = {};
          if (_.has(answer, 'tolerance')) {
            responseParam = {
              responseparam: {
                '@_type': 'tolerance',
                '@_default': _.get(answer, 'tolerance', 0),
              },
            };
          }
          answerObject = {
            '@_answer': answer.title,
            ...responseParam,
            ...correcthint,
          };
        } else if (answer.correct && firstCorrectAnswerParsed) {
          additionalAnswers.push({
            '@_answer': answer.title,
            ...correcthint,
          });
        }
      }
    });
    answerObject = {
      ...answerObject,
      additional_answer: additionalAnswers,
      formulaequationinput: {
        '#text': '',
      },
    };
    return answerObject;
  }

  getAnswerHints(elementObject) {
    const feedback = elementObject?.feedback;
    let correcthint = {};
    if (feedback !== undefined && feedback !== '') {
      correcthint = {
        correcthint: {
          '#text': feedback,
        },
      };
    }
    return correcthint;
  }

  hasAttributeWithValue(obj, attr) {
    return _.has(obj, attr) && _.get(obj, attr, '').trim() !== '';
  }

  buildOLX() {
    const { problemType } = this.problemState;
    let problemString = '';
    switch (problemType) {
      case ProblemTypeKeys.MULTISELECT:
        problemString = this.buildMultiSelectProblem(ProblemTypeKeys.MULTISELECT, 'checkboxgroup', 'choice');
        break;
      case ProblemTypeKeys.DROPDOWN:
        problemString = this.buildMultiSelectProblem(ProblemTypeKeys.DROPDOWN, 'optioninput', 'option');
        break;
      case ProblemTypeKeys.SINGLESELECT:
        problemString = this.buildMultiSelectProblem(ProblemTypeKeys.SINGLESELECT, 'choicegroup', 'choice');
        break;
      case ProblemTypeKeys.TEXTINPUT:
        problemString = this.buildTextInput();
        break;
      case ProblemTypeKeys.NUMERIC:
        problemString = this.buildNumericInput();
        break;
      default:
        break;
    }
    return problemString;
  }
}

export default ReactStateOLXParser;
