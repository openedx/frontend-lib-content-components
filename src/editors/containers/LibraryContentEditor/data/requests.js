import { StrictDict } from '../../../utils';
import { RequestKeys } from '../../../data/constants/requests';
import { networkRequest } from '../../../data/redux/thunkActions/requests';

/* eslint-disable import/no-cycle */
import { selectors } from '../../../data/redux';
import api from './api';
import apiMethods from '../../../data/services/cms/api';

/**
 * Tracked fetchV1Libraries api method.
 * Tracked to the `fetchV1Libraries` request key.
 * @param {[func]} onSuccess - onSuccess method ((response) => { ... })
 * @param {[func]} onFailure - onFailure method ((error) => { ... })
 */
export const fetchV1Libraries = ({ ...rest }) => (dispatch, getState) => {
  dispatch(networkRequest({
    requestKey: RequestKeys.fetchV1Libraries,
    promise: api.fetchV1Libraries({
      studioEndpointUrl: selectors.app.studioEndpointUrl(getState()),
    }),
    ...rest,
  }));
};

/**
 * Tracked fetchV2Libraries api method.
 * Tracked to the `fetchV2Libraries` request key.
 * @param {[func]} onSuccess - onSuccess method ((response) => { ... })
 * @param {[func]} onFailure - onFailure method ((error) => { ... })
 */
export const fetchV2Libraries = ({ ...rest }) => (dispatch, getState) => {
  dispatch(networkRequest({
    requestKey: RequestKeys.fetchV2Libraries,
    promise: api.fetchV2Libraries({
      studioEndpointUrl: selectors.app.studioEndpointUrl(getState()),
    }),
    ...rest,
  }));
};

/**
 * Tracked fetchV2LibraryContent api method.
 * Tracked to the `fetchV2LibraryContent` request key.
 * @param {[func]} onSuccess - onSuccess method ((response) => { ... })
 * @param {[func]} onFailure - onFailure method ((error) => { ... })
 */
export const fetchV1LibraryContent = ({ libraryId, ...rest }) => (dispatch, getState) => {
  dispatch(networkRequest({
    requestKey: RequestKeys.fetchV1LibraryContent,
    promise: api.fetchV1LibraryContent({
      studioEndpointUrl: selectors.app.studioEndpointUrl(getState()),
      libraryId,
    }),
    ...rest,
  }));
};

/**
 * Tracked fetchV2LibraryContent api method.
 * Tracked to the `fetchV2LibraryContent` request key.
 * @param {[func]} onSuccess - onSuccess method ((response) => { ... })
 * @param {[func]} onFailure - onFailure method ((error) => { ... })
 */
export const fetchV2LibraryContent = ({ libraryId, ...rest }) => (dispatch, getState) => {
  dispatch(networkRequest({
    requestKey: RequestKeys.fetchV2LibraryContent,
    promise: api.fetchV2LibraryContent({
      studioEndpointUrl: selectors.app.studioEndpointUrl(getState()),
      libraryId,
    }),
    ...rest,
  }));
};

/**
 * Tracked fetchV1LibraryBlock api method.
 * Tracked to the `fetchV1LibraryBlock` request key.
 * @param {[func]} onSuccess - onSuccess method ((response) => { ... })
 * @param {[func]} onFailure - onFailure method ((error) => { ... })
 */
export const fetchV1LibraryBlock = ({ blockId, ...rest }) => (dispatch, getState) => {
  dispatch(networkRequest({
    requestKey: RequestKeys.fetchV1LibraryBlock,
    promise: apiMethods.fetchBlockById({
      studioEndpointUrl: selectors.app.studioEndpointUrl(getState()),
      blockId,
    }),
    ...rest,
  }));
};

/**
 * Tracked fetchChildrenInfo api method.
 * Tracked to the `fetchChildrenInfo` request key.
 * @param {[func]} onSuccess - onSuccess method ((response) => { ... })
 * @param {[func]} onFailure - onFailure method ((error) => { ... })
 */
export const fetchChildrenInfo = ({ ...rest }) => (dispatch, getState) => {
  dispatch(networkRequest({
    requestKey: RequestKeys.fetchChildrenInfo,
    promise: api.fetchChildrenInfo({
      studioEndpointUrl: selectors.app.studioEndpointUrl(getState()),
      blockId: selectors.app.blockId(getState()),
    }),
    ...rest,
  }));
};

export default StrictDict({
  fetchV1Libraries,
  fetchV2Libraries,
  fetchV1LibraryContent,
  fetchV2LibraryContent,
  fetchV1LibraryBlock,
  fetchChildrenInfo,
});
