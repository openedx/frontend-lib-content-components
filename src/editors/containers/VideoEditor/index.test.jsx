import React from 'react';
import { shallow } from '@edx/react-unit-test-utils';

import { formatMessage } from '../../../testUtils';
import { selectors } from '../../data/redux';
import { RequestKeys } from '../../data/constants/requests';
import { VideoEditor, mapStateToProps, mapDispatchToProps } from '.';

jest.mock('../EditorContainer', () => 'EditorContainer');
jest.mock('./components/VideoEditorModal', () => 'VideoEditorModal');

jest.mock('./hooks', () => ({
  ErrorContext: {
    Provider: 'ErrorContext.Provider',
  },
  errorsHook: jest.fn(() => ({
    error: 'hooks.errorsHook.error',
    validateEntry: jest.fn().mockName('validateEntry'),
  })),
  fetchVideoContent: jest.fn().mockName('fetchVideoContent'),
}));

jest.mock('../../data/redux', () => ({
  selectors: {
    requests: {
      isFinished: jest.fn((state, params) => ({ isFailed: { state, params } })),
    },
    app: {
      isLibrary: jest.fn(state => ({ isLibrary: state })),
    },
  },
}));

jest.mock('@edx/paragon', () => ({
  ...jest.requireActual('@edx/paragon'),
  Spinner: 'Spinner',
}));

describe('VideoEditor', () => {
  const props = {
    onClose: jest.fn().mockName('props.onClose'),
    intl: { formatMessage },
    studioViewFinished: false,
    isLibrary: false,
  };
  describe('snapshots', () => {
    test('renders as expected with default behavior', () => {
      expect(shallow(<VideoEditor {...props} />).snapshot).toMatchSnapshot();
    });
    test('renders as expected with default behavior', () => {
      expect(shallow(<VideoEditor {...props} studioViewFinished />).snapshot).toMatchSnapshot();
    });
  });
  describe('mapStateToProps', () => {
    const testState = { A: 'pple', B: 'anana', C: 'ucumber' };
    test('studioViewFinished from requests.isFinished', () => {
      expect(
        mapStateToProps(testState).studioViewFinished,
      ).toEqual(selectors.requests.isFinished(testState, { requestKey: RequestKeys.fetchStudioView }));
    });
    test('isLibrary from app.isLibrary', () => {
      expect(
        mapStateToProps(testState).isLibrary,
      ).toEqual(selectors.app.isLibrary(testState));
    });
  });
  describe('mapDispatchToProps', () => {
    test('is empty', () => {
      expect(mapDispatchToProps).toEqual({});
    });
  });
});
