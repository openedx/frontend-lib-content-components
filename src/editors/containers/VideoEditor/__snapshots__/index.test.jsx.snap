// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoEditor snapshots renders as expected with default behavior 1`] = `
<ErrorContext.Provider
  value="hooks.errorsHook.error"
>
  <EditorContainer
    onClose={[MockFunction props.onClose]}
    returnFunction={null}
    validateEntry={[MockFunction validateEntry]}
  >
    <div
      style={
        Object {
          "left": "50%",
          "position": "absolute",
          "top": "50%",
          "transform": "translate(-50%, -50%)",
        }
      }
    >
      <Spinner
        animation="border"
        className="m-3"
        screenreadertext="loading"
      />
    </div>
  </EditorContainer>
</ErrorContext.Provider>
`;

exports[`VideoEditor snapshots renders as expected with default behavior 2`] = `
<ErrorContext.Provider
  value="hooks.errorsHook.error"
>
  <EditorContainer
    onClose={[MockFunction props.onClose]}
    returnFunction={null}
    validateEntry={[MockFunction validateEntry]}
  >
    <div
      className="video-editor"
    >
      <VideoEditorModal
        isLibrary={false}
      />
    </div>
  </EditorContainer>
</ErrorContext.Provider>
`;
